Addrmap iterator

krcb197 opened the pull request:
Closes #47

krcb197 commented on src/peakrdl_python/templates/peakrdl_python/base.py:
```diff
@@ -36,8 +39,28 @@ def inst_name(self) -> str:
+        return_name = self.inst_name
+        p = self.parent
+        while p is not None :
+            return_name = p.inst_name + "." + return_name
```
@mpriestleyidex 

If this function has to provide the "full" name surely this would be p.full_name + '.' self.inst_name, otherwise with will only go one layer back

mpriestleyidex replied:
Good point....

krcb197 replied:
I have made this change

krcb197 commented on src/peakrdl_python/templates/peakrdl_python/base.py:
```diff
@@ -36,8 +39,28 @@ def inst_name(self) -> str:
+        return self.__parent
+
+    @property
+    def full_name(self) -> str:
```
I would favour changing this to `full_inst_name` to make is more consistent with the `inst_name` above

krcb197 replied:
I have made this change

krcb197 commented on src/peakrdl_python/templates/peakrdl_python/memory.py:
```diff
@@ -4,13 +4,14 @@
 
 
-class Memory(Node):
+class Memory(Node, ABC):
```
Now that I have fixed the bug that memories that allows memories to have registers within them. I think the 

```{{ child_getter(node, "get_registers",   "Reg",        systemrdlRegNode) }}`` is needed in here

krcb197 replied:
I have made a separate `readable_registers` and `writeable_registers` getter.

krcb197 replied:
I have moved the readable and writeable register to `reg_definition.py.jinja` to allow its use in the memories

krcb197 commented on src/peakrdl_python/templates/addrmap.py.jinja:
```diff
@@ -155,6 +163,34 @@ class {{get_fully_qualified_type_name(node)}}_cls(AddressMap):
             {%- endif %}
         {% endfor %}
+
+    {{ child_getter(node, "get_registers",   "Reg",        systemrdlRegNode) }}
```
@mpriestleyidex 
The field version has a separate interator for readable_registers and writable_registers. When striding an address map I often want to operate on these separately depending on what is happening

mpriestleyidex replied:
I agree - the generic dumping / checking functions I have in mind are normally looking for readable registers.

krcb197 replied:
I have made a separate `readable_registers` and `writeable_registers` getter.

krcb197 reviewed (commented):
@mpriestleyidex 

Thank you for your submission, I have made some comments please let me know what you think.

mpriestleyidex commented on src/peakrdl_python/templates/addrmap.py.jinja:
```diff
@@ -155,6 +163,34 @@ class {{get_fully_qualified_type_name(node)}}_cls(AddressMap):
+    {{ child_getter(node, "get_registers",   "Reg",        systemrdlRegNode) }}
+    {{ child_getter(node, "get_memories",    "Memory",     systemrdlMemNode) }}
+    {{ child_getter(node, "get_addressmaps", "AddressMap", systemrdlAddrmapNode) }}
+    {{ child_getter(node, "get_regfiles",    "RegFile",    systemrdlRegfileNode) }}
```
I've not got my head around what an RDL regfile is for, and why I'd choose to use one over an addrmap. So I tend to forget about them. I'm happy to merge into 'sections'.

krcb197 replied:
The SystemRDL specification gives a clue:

> 13.1 Introduction
During generation, the address map can be converted into an HDL module. All registers and fields
instantiated within an address map file shall be generated within this module. Therefore, some properties are
only valid for addrmaps and not for regfiles. Other than these properties and their suggested behavior, there
is no difference between address maps and register files.

My understanding is you use AddrMap to guide any generator that it may put things into separate modules, whereas RegFile grouping concept that doesn't infer separation of the generated outputs. At the level of the PythonAPI (once which is a generated output) these two concepts can be merged.

mpriestleyidex replied:
Thanks, that makes sense, and I can see where I'd want to use them now.
Yes, let's have a single iterator.

krcb197 replied:
This was quite easy to do with the macro you already provided:

Address Map Class changed to:
```
{{ child_getter(node, "get_sections", "Union[AddressMap, RegFile]", (systemrdlAddrmapNode, systemrdlRegfileNode)) }}
```

RegFile Class changed to: 
```
{{ child_getter(node, "get_sections",  "RegFile", systemrdlRegfileNode) }}
```

mpriestleyidex commented on src/peakrdl_python/templates/addrmap.py.jinja:
```diff
@@ -155,6 +163,34 @@ class {{get_fully_qualified_type_name(node)}}_cls(AddressMap):
+{%- endmacro %}
+
+{%- macro child_getter(node, getter_name, child_type, child_rdltype) %}
+    def {{getter_name}}(self, unroll=False) -> Iterable[Union[{{child_type}},Tuple[{{child_type}},...]]]:
```
Hmm. I think there needs to be some means of discovering and getting hold of an array of things. It could be handled by `Node`s having an `is_array_member` property and then a `get_enclosing_array`.

Or we could change the default of `unroll` to `True`, and provide the property too by:
```
    registers = property(fget=get_registers)
```

krcb197 replied:
Unlike the system RDL, in the python API the array has no instance, each element of the array is a separate instance within its parent. 

What is the use case you have in mind for accessing the array but not its individual elements?

krcb197 replied:
I have started trying to write some examples:
- Dumping all readable registers to a json file (having first put them in a list/dictionary structure)
- Resetting all the writable registers to their default values.

I am starting to appreciate the issue you have been talking about. It is highlighting a few other practical issues. I will add these examples into the documentation as part of this branch so you can take a look.

mpriestleyidex replied:
Sorry, I've not been keeping up for the last couple of days (in which you seem to have done an awful lot!).
Yes, dumping to JSON is a good example of where you'd want to treat arrays as arrays.

krcb197 replied:
There are two new examples under `example/tranversing_address_map`. I think these address the points you were trying to make. Please let me know what you think, I need to some test cases to check out the generators before we merge this pull request.

krcb197 commented on src/peakrdl_python/templates/addrmap.py.jinja:
```diff
@@ -155,6 +163,33 @@ class {{get_fully_qualified_type_name(node)}}_cls(AddressMap):
+{%- endmacro %}
+
+{%- macro child_getter(node, getter_name, child_type, child_rdltype) %}
+    def {{getter_name}}(self, unroll=False) -> Iterable[Union[{{child_type}},Tuple[{{child_type}},...]]]:
```
I think this should be type hinted as `Iterator ` rather than `Iterable` See[mypy type hints](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

krcb197 replied:
The python documentation for typing a generator, shows the use of `Iterator`. I have gone ahead and changed this.

krcb197 commented:
@mpriestleyidex 

I have added a new example for writing the register based on a JSON file which only write the non-null values

I have added a set of tests for the new iterators, so I think this Pull Request is ready to deployed. Please let me know what you think.
